plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/avro/PulsarAvroRowDecoderFactory.java

```diff
@@ -0,0 +1,196 @@
+import org.apache.pulsar.common.schema.SchemaInfo;
+
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+import static io.trino.spi.type.DateType.DATE;
+import static io.trino.spi.type.TimeType.TIME_MILLIS;
+import static io.trino.spi.type.TimestampType.TIMESTAMP_MILLIS;
+import static io.trino.spi.type.VarcharType.createUnboundedVarcharType;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * PulsarRowDecoderFactory for {@link org.apache.pulsar.shade.org.apache.pulsar.common.schema.SchemaType#AVRO}.
+ */
+public class PulsarAvroRowDecoderFactory
+        implements PulsarRowDecoderFactory
+{
+    private TypeManager typeManager;
+
+    public PulsarAvroRowDecoderFactory(TypeManager typeManager)
+    {
+        this.typeManager = typeManager;
+    }
+
+    @Override
+    public PulsarRowDecoder createRowDecoder(TopicName topicName,
+                                             SchemaInfo schemaInfo,
+                                             Set<DecoderColumnHandle> columns)
+    {
+        return new PulsarAvroRowDecoder((GenericAvroSchema) GenericAvroSchema.of(schemaInfo), columns);
+    }
+
+    @Override
+    public List<ColumnMetadata> extractColumnMetadata(TopicName topicName,
+                                                      SchemaInfo schemaInfo,
+                                                      PulsarColumnHandle.HandleKeyValueType handleKeyValueType)
+    {
+        List<ColumnMetadata> columnMetadata;
+        String schemaJson = new String(schemaInfo.getSchema(), StandardCharsets.ISO_8859_1);
+        if (Strings.nullToEmpty(schemaJson).trim().isEmpty()) {
+            throw new TrinoException(NOT_SUPPORTED, "Topic "
+                    + topicName.toString() + " does not have a valid schema");
+        }
+        Schema schema;
+        try {
+            schema = GenericJsonSchema.of(schemaInfo).getAvroSchema();
+        }
+        catch (SchemaParseException ex) {
+            throw new TrinoException(NOT_SUPPORTED, "Topic "
+                    + topicName.toString() + " does not have a valid schema");
+        }
+
+        try {
+            columnMetadata = schema.getFields().stream()
+                    .map(field ->
+                            new PulsarColumnMetadata(PulsarColumnMetadata.getColumnName(handleKeyValueType, field.name()), parseAvroTrinoType(
+                                    field.name(), field.schema()), field.schema().toString(), null, false, false,
+                                    handleKeyValueType, new PulsarColumnMetadata.DecoderExtraInfo(field.name(),
+                                    null, null))
+
+                    ).collect(toList());
+        }
+        catch (StackOverflowError e) {
```

Comment:
Why it can happen?

Reply:
If a avro schema contains a field referring itself, it could possibly cause stackoverflow when trying to extract column matadata from the schema. I think it's more like a safety net.